OpenRA.Mods.Common/FileFormats/WavReader.cs

```diff
@@ -112,69 +117,86 @@ public static float WaveLength(Stream s)
+			readonly int[] index;
+
+			readonly byte[] interleaveBuffer;
+			int outOffset;
+			int currentBlock;
+
+			public WavStream(Stream stream, int dataSize, short blockAlign, short channels, int uncompressedSize) : base(stream)
+			{
+				this.channels = channels;
+				numBlocks = dataSize / blockAlign;
+				blockDataSize = blockAlign - (channels * 4);
+				outputSize = uncompressedSize * channels * 2;
+				predictor = new int[channels];
+				index = new int[channels];
+
+				interleaveBuffer = new byte[channels * 16];
+			}
 
-			// Decode each block of IMA ADPCM data in RawOutput
-			for (var block = 0; block < numBlocks; block++)
+			protected override bool BufferData(Stream baseStream, Queue<byte> data)
 			{
+				// Decode each block of IMA ADPCM data
 				// Each block starts with a initial state per-channel
-				for (var c = 0; c < Channels; c++)
+				for (var c = 0; c < channels; c++)
 				{
-					predictor[c] = s.ReadInt16();
-					index[c] = s.ReadUInt8();
-					/* unknown/reserved */ s.ReadUInt8();
+					predictor[c] = baseStream.ReadInt16();
+					index[c] = baseStream.ReadUInt8();
+					baseStream.ReadUInt8(); // Unknown/Reserved
 
 					// Output first sample from input
-					output[outOffset++] = (byte)predictor[c];
-					output[outOffset++] = (byte)(predictor[c] >> 8);
+					data.Enqueue((byte)predictor[c]);
+					data.Enqueue((byte)(predictor[c] >> 8));
+					outOffset += 2;
 
 					if (outOffset >= outputSize)
-						return output;
+						return true;
 				}
 
 				// Decode and output remaining data in this block
 				var blockOffset = 0;
 				while (blockOffset < blockDataSize)
 				{
-					for (var c = 0; c < Channels; c++)
+					for (var c = 0; c < channels; c++)
 					{
 						// Decode 4 bytes (to 16 bytes of output) per channel
-						var chunk = s.ReadBytes(4);
+						var chunk = baseStream.ReadBytes(4);
 						var decoded = ImaAdpcmReader.LoadImaAdpcmSound(chunk, ref index[c], ref predictor[c]);
 
 						// Interleave output, one sample per channel
-						var outOffsetChannel = outOffset + (2 * c);
+						var interleaveChannelOffset = 2 * c;
 						for (var i = 0; i < decoded.Length; i += 2)
 						{
-							var outOffsetSample = outOffsetChannel + i;
-							if (outOffsetSample >= outputSize)
-								return output;
-
-							output[outOffsetSample] = decoded[i];
-							output[outOffsetSample + 1] = decoded[i + 1];
-							outOffsetChannel += 2 * (Channels - 1);
+							var interleaveSampleOffset = interleaveChannelOffset + i;
+							interleaveBuffer[interleaveSampleOffset] = decoded[i];
+							interleaveBuffer[interleaveSampleOffset + 1] = decoded[i + 1];
+							interleaveChannelOffset += 2 * (channels - 1);
 						}
 
 						blockOffset += 4;
 					}
 
-					outOffset += 16 * Channels;
+					var outputRemaining = outputSize - outOffset;
+					var toCopy = Math.Min(outputRemaining, interleaveBuffer.Length);
+					for (var i = 0; i < toCopy; i++)
+						data.Enqueue(interleaveBuffer[i]);
```

Comment:
Instead of bailing out mid-decoding of channels, we decode them all, but then only return as much data as expected.

This represents a bug-fix: Previously the early bail out would mean some bytes right at the end would not be decoded and would be left with their default value of 0. Now we decode these and set their values correctly.

Both the old and new methods output the same number of bytes, it's just the left few bytes in the old method would often be mistakenly zeroed.